Port XT PR #403: [RPC][Bitcoin-TX] Add support for sequence number

sickpig opened the pull request:
bitcoin/bitcoin#7957 - [RPC][Bitcoin-TX] Add support for sequence number
bitcoin/bitcoin#8164 - [Bitcoin-Tx] fix missing test fixtures, fix 32bit atoi issue
bitcoin/bitcoin#8171 - [RPC] Fix createrawtx sequence number unsigned int parsing

[RPC][Bitcoin-TX] Add support for sequence number

dgenr8 committed:
Port XT PR #403: [RPC][Bitcoin-TX] Add support for sequence number

bitcoin/bitcoin#7957 - [RPC][Bitcoin-TX] Add support for sequence number
bitcoin/bitcoin#8164 - [Bitcoin-Tx] fix missing test fixtures, fix 32bit atoi issue
bitcoin/bitcoin#8171 - [RPC] Fix createrawtx sequence number unsigned int parsing

[RPC][Bitcoin-TX] Add support for sequence number

gandrewstone commented on src/rpc/rawtransaction.cpp:
```diff
@@ -415,6 +417,18 @@ UniValue createrawtransaction(const UniValue &params, bool fHelp)
+            else
+                nSequence = (uint32_t)seqNr64;
+        }
+
```
if its not a number we should throw a JSONRPCError, not just silently ignore it.

EDIT: or better first use  http://en.cppreference.com/w/cpp/string/basic_string/stoul to attempt to convert the string to a number, and only throw JSONRPCError if stoul() fails

sickpig replied:
Yeah you have point. Going to introduce these changes.

sickpig replied:
On second thought why not just rising an error if `sequence` is not a number? 

thing is that if someone is trying to set something like `-1` as `sequence`, `stoull` will return `18446744073709551615`. I don't think this is something that we wanted to accept, no?

gandrewstone replied:
My point was what if someone used "25" as the sequence, not 25 -- that is, for some json translation reason the number is being sent as a string.  There is a tension in the RPC interface between end-user (command-line and scripts) convenience and programmatic correctness (where really we can insist that 25 be 25 not "25".  Since end-user convenience doesn't much impact programmatic use case (except possibly by hiding bugs), I tend to favor convenience.  But you are right that that can hide bugs like -1 in the input, unless we check for them.

But we aren't going to solve this today,  For example, the RPC interface only wants 1 or 0 for true/false values which is pretty awkward.  So I am ok with this being a rigid, integer-only parameter.

gandrewstone commented on qa/rpc-tests/rawtransactions.py:
```diff
@@ -238,6 +238,27 @@ def run_test(self):
+        inputs  = [ {'txid' : "1d1d4e24ed99057e84c3f80fd8fbec79ed9e1acee37da269356ecea000000000", 'vout' : 1, 'sequence' : 4294967296}]
+        outputs = { self.nodes[0].getnewaddress() : 1 }
+        assert_raises(JSONRPCException, self.nodes[0].createrawtransaction, inputs, outputs)
+
```
add check for 'sequence' : "some invalid string"

sickpig committed:
Raise an error in case provided sequence number is not a number.

Add also a check to test this new behavior, i.e. trying to pass a string
as sequence number and check that an error is raised properly.

sickpig commented:
- rebased

- raise an error in case sequence param is not a number

- add a test for not-a-number sequence parameter
